engine/src/test/java/org/camunda/bpm/engine/test/bpmn/tasklistener/TaskListenerEventLifecycleTest.java

```diff
@@ -0,0 +1,586 @@
+    // given
+    BpmnModelInstance model = createModelWithTaskEventsRecorderOnAssignedUserTask(TRACKED_EVENTS,
+                                                                                  null,
+                                                                                  TaskListener.EVENTNAME_COMPLETE,
+                                                                                  AssigneeAssignment.class);
+    testRule.deploy(model);
+    runtimeService.startProcessInstanceByKey("process");
+    Task task = taskService.createTaskQuery().singleResult();
+
+    // when
+    taskService.complete(task.getId());
+
+    // then
+    LinkedList<String> orderedEvents = RecorderTaskListener.getOrderedEvents();
+
+    assertThat(orderedEvents.size()).isEqualTo(2);
+    assertThat(orderedEvents.getLast()).isEqualToIgnoringCase(TaskListener.EVENTNAME_COMPLETE);
+  }
+
+  @Test
+  public void shouldNotFireDeleteEventOnTaskDeletedInCompleteListener() {
+    // given
+    thrown.expectMessage("The task cannot be deleted because is part of a running process");
+    BpmnModelInstance model = createModelWithTaskEventsRecorderOnAssignedUserTask(TRACKED_EVENTS,
+                                                                                  null,
+                                                                                  TaskListener.EVENTNAME_COMPLETE,
+                                                                                  TaskDeleteTaskListener.class);
+    testRule.deploy(model);
+    runtimeService.startProcessInstanceByKey("process");
+    Task task = taskService.createTaskQuery().singleResult();
+
+    // when
+    taskService.complete(task.getId());
+
+    // then
+    LinkedList<String> orderedEvents = RecorderTaskListener.getOrderedEvents();
+
+    assertThat(orderedEvents.size()).isEqualTo(2);
+    assertThat(orderedEvents.getLast()).isEqualToIgnoringCase(TaskListener.EVENTNAME_COMPLETE);
+  }
+
+  @Test
+  public void shouldNotFireCompleteEventOnTaskCompletedInCompleteListener() {
+    // given
+    thrown.expectMessage("invalid task state");
+    BpmnModelInstance model = createModelWithTaskEventsRecorderOnAssignedUserTask(TRACKED_EVENTS,
+                                                                                  null,
+                                                                                  TaskListener.EVENTNAME_COMPLETE,
+                                                                                  CompletingTaskListener.class);
+    testRule.deploy(model);
+    runtimeService.startProcessInstanceByKey("process");
+    Task task = taskService.createTaskQuery().singleResult();
+
+    // when
+    taskService.complete(task.getId());
+
+    // then
+    LinkedList<String> orderedEvents = RecorderTaskListener.getOrderedEvents();
+
+    assertThat(orderedEvents.size()).isEqualTo(1);
+    assertThat(orderedEvents.getFirst()).isEqualToIgnoringCase(TaskListener.EVENTNAME_CREATE);
+  }
```

Comment:
When I try do to a deletion in the complete listener, the COMPLETE event is fired.
When I try do to another completion in the complete listener, the COMPLETE event is **NOT** fired.
That looks a bit confusing to me.

Reply:
The COMPLETE event should be fired in the second test. It's actually wrong to expect only a single event, I'll look into it.